File: frontend/src/types/ai.ts
export type AIProviderType = 'gemini' | 'openai' | 'ollama';

export interface AIProviderConfig {
  id: string;
  name: string;
  provider_type: AIProviderType;
  enabled: boolean;
  active: boolean;
  base_url: string | null;
  api_key_configured: boolean;
  api_key_display: string;
  api_key_env_var: string | null;
  model_name: string;
  vision_enabled: boolean;
  timeout_seconds: number;
  max_output_tokens: number | null;
  temperature: number | null;
  last_test_datetime: string | null;
  last_test_status: string | null;
  last_error_message: string | null;
  created_datetime: string;
  updated_datetime: string | null;
}

export interface ListAIProvidersResponse {
  providers: AIProviderConfig[];
}

export interface GetAIProviderResponse {
  provider: AIProviderConfig;
}

export interface CreateAIProviderInput {
  name: string;
  provider_type: AIProviderType;
  enabled?: boolean;
  active?: boolean;
  base_url?: string | null;
  api_key_value?: string | null;
  api_key_env_var?: string | null;
  model_name: string;
  vision_enabled?: boolean;
  timeout_seconds?: number;
  max_output_tokens?: number | null;
  temperature?: number | null;
}

export interface UpdateAIProviderInput {
  name?: string | null;
  provider_type?: AIProviderType | null;
  enabled?: boolean | null;
  active?: boolean | null;
  base_url?: string | null;
  api_key_value?: string | null;
  api_key_env_var?: string | null;
  model_name?: string | null;
  vision_enabled?: boolean | null;
  timeout_seconds?: number | null;
  max_output_tokens?: number | null;
  temperature?: number | null;
  clear_api_key?: boolean | null;
}

export interface DeleteAIProviderResponse {
  provider_id: string;
  deleted: boolean;
}

export interface AIProviderTestResult {
  provider_id: string;
  provider_type: AIProviderType;
  model_name: string;
  status: 'success' | 'failed';
  message: string;
  tested_datetime: string;
}

export interface AISettings {
  ai_assist_enabled: boolean;
  active_provider_id: string | null;
  active_provider_name: string | null;
  active_provider_type: AIProviderType | null;
  active_model_name: string | null;
}

export interface GetAISettingsResponse {
  settings: AISettings;
}

export interface UpdateAISettingsInput {
  ai_assist_enabled?: boolean | null;
  active_provider_id?: string | null;
}

File: api/internal/models/ai.go
package models

import (
	"encoding/json"
	"time"
)

type OptionalBool struct {
	Set   bool
	Value *bool
}

func (o *OptionalBool) UnmarshalJSON(data []byte) error {
	o.Set = true

	if string(data) == "null" {
		o.Value = nil
		return nil
	}

	var value bool
	if err := json.Unmarshal(data, &value); err != nil {
		return err
	}

	o.Value = &value
	return nil
}

type OptionalInt struct {
	Set   bool
	Value *int
}

func (o *OptionalInt) UnmarshalJSON(data []byte) error {
	o.Set = true

	if string(data) == "null" {
		o.Value = nil
		return nil
	}

	var value int
	if err := json.Unmarshal(data, &value); err != nil {
		return err
	}

	o.Value = &value
	return nil
}

type OptionalFloat64 struct {
	Set   bool
	Value *float64
}

func (o *OptionalFloat64) UnmarshalJSON(data []byte) error {
	o.Set = true

	if string(data) == "null" {
		o.Value = nil
		return nil
	}

	var value float64
	if err := json.Unmarshal(data, &value); err != nil {
		return err
	}

	o.Value = &value
	return nil
}

type AIProviderConfig struct {
	ID               string     `json:"id"`
	Name             string     `json:"name"`
	ProviderType     string     `json:"provider_type"`
	Enabled          bool       `json:"enabled"`
	Active           bool       `json:"active"`
	BaseURL          *string    `json:"base_url"`
	APIKeyConfigured bool       `json:"api_key_configured"`
	APIKeyDisplay    string     `json:"api_key_display"`
	APIKeyEnvVar     *string    `json:"api_key_env_var"`
	ModelName        string     `json:"model_name"`
	VisionEnabled    bool       `json:"vision_enabled"`
	TimeoutSeconds   int        `json:"timeout_seconds"`
	MaxOutputTokens  *int       `json:"max_output_tokens"`
	Temperature      *float64   `json:"temperature"`
	LastTestDatetime *time.Time `json:"last_test_datetime"`
	LastTestStatus   *string    `json:"last_test_status"`
	LastErrorMessage *string    `json:"last_error_message"`
	CreatedDatetime  time.Time  `json:"created_datetime"`
	UpdatedDatetime  *time.Time `json:"updated_datetime"`
}

type ListAIProvidersResponse struct {
	Providers []AIProviderConfig `json:"providers"`
}

type GetAIProviderResponse struct {
	Provider AIProviderConfig `json:"provider"`
}

type CreateAIProviderRequest struct {
	Name            string   `json:"name"`
	ProviderType    string   `json:"provider_type"`
	Enabled         *bool    `json:"enabled"`
	Active          *bool    `json:"active"`
	BaseURL         *string  `json:"base_url"`
	APIKeyValue     *string  `json:"api_key_value"`
	APIKeyEnvVar    *string  `json:"api_key_env_var"`
	ModelName       string   `json:"model_name"`
	VisionEnabled   *bool    `json:"vision_enabled"`
	TimeoutSeconds  *int     `json:"timeout_seconds"`
	MaxOutputTokens *int     `json:"max_output_tokens"`
	Temperature     *float64 `json:"temperature"`
}

type PatchAIProviderRequest struct {
	Name            OptionalString  `json:"name"`
	ProviderType    OptionalString  `json:"provider_type"`
	Enabled         OptionalBool    `json:"enabled"`
	Active          OptionalBool    `json:"active"`
	BaseURL         OptionalString  `json:"base_url"`
	APIKeyValue     OptionalString  `json:"api_key_value"`
	APIKeyEnvVar    OptionalString  `json:"api_key_env_var"`
	ModelName       OptionalString  `json:"model_name"`
	VisionEnabled   OptionalBool    `json:"vision_enabled"`
	TimeoutSeconds  OptionalInt     `json:"timeout_seconds"`
	MaxOutputTokens OptionalInt     `json:"max_output_tokens"`
	Temperature     OptionalFloat64 `json:"temperature"`
	ClearAPIKey     OptionalBool    `json:"clear_api_key"`
}

type DeleteAIProviderResponse struct {
	ProviderID string `json:"provider_id"`
	Deleted    bool   `json:"deleted"`
}

type AIProviderTestResult struct {
	ProviderID     string    `json:"provider_id"`
	ProviderType   string    `json:"provider_type"`
	ModelName      string    `json:"model_name"`
	Status         string    `json:"status"`
	Message        string    `json:"message"`
	TestedDatetime time.Time `json:"tested_datetime"`
}

type AISettings struct {
	AIAssistEnabled    bool    `json:"ai_assist_enabled"`
	ActiveProviderID   *string `json:"active_provider_id"`
	ActiveProviderName *string `json:"active_provider_name"`
	ActiveProviderType *string `json:"active_provider_type"`
	ActiveModelName    *string `json:"active_model_name"`
}

type GetAISettingsResponse struct {
	Settings AISettings `json:"settings"`
}

type PatchAISettingsRequest struct {
	AIAssistEnabled  OptionalBool   `json:"ai_assist_enabled"`
	ActiveProviderID OptionalString `json:"active_provider_id"`
}
